### PR TITLE
Add backwards compatibility flag for session-key fix.

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 
Follow-on fix from PR #2024, which got merged before fixing the compatibility-breaking change, i.e. that the session window written as part of the record key will now contain the start _and_ end time, where previously it was just the start time. 

This PR addresses this by adding a new `windowed.session.key.legacy` setting. Users can set this to `true` to use the legacy key format. More importantly, this setting will automatically be `true` for all existing queries, ensuring backwards compatibility.

*Question*: Is my change in `SchemaKGroupedStream` sufficient to ensure all queries use the old format and does this PR address all the compatibility concerns of PR #2024?  Answers on a postcard...

### Testing done 
Added tests to session-window.json to cover test case.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by hjafarpour** [comment]
@big-andy-coates @rodesai and I raised some issues about the previous PR (#2024) which were not addressed: https://github.com/confluentinc/ksql/pull/2024#discussion_r224178294
I think we should reopen the previous PR first and resolve them before this PR.

**Comment by big-andy-coates** [comment]
Apologies @hjafarpour - I missed that comment when checking over the PR before merging.

File: ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java [hunk 1]
```diff
@@ -104,6 +104,15 @@
       + " VARCHAR SUBSTRING(str VARCHAR, pos INT, length INT), where pos is base-one indexed,"
       + " and the last argument is the length of the substring to extract.";
 
+  public static final String KSQL_WINDOWED_SESSION_KEY_LEGACY_CONFIG =
+      KSQL_CONFIG_PROPERTY_PREFIX + "windowed.session.key.legacy";
+
+  private static final String KSQL_WINDOWED_SESSION_KEY_LEGACY_DOC = ""
+      + "Version 5.1 of KSQL and earlier incorrectly excluded the end time in the record key in "
+      + "Kafka for session windowed data. Setting this value to true will make KSQL expect and "
+      + "continue to store session keys without the end time. With the default value of true "
```
Comment (hjafarpour) [inline ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java]: Default value is `false`! It's true for the existing queries only.

**Comment by big-andy-coates** [comment]
@hjafarpour 

> In addition to SchemaKGroupedStream.java, we define the window serde in SchemaKTable.java and StructuredDataSourceNode.java. So you should check for the legacy session window there too.

I don't believe anything needs to change in these classes. If you think I've missed something then please provide a test case that highlights the issue.

File: ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java (hunk 1 above)
Comment (big-andy-coates) [reply ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java]: done

**Comment by hjafarpour** [comment]
Yes, you are correct. The following line from `StructuredDataSourceNode.java` will use the proper end time regardless if it is `-` or actual timestamp: 

`final String end = window instanceof SessionWindow ? String.valueOf(window.end()) : "-";`

You only need to check the flag when you set the serde to write data into topic.

**Comment by big-andy-coates** [comment]
Hi @hojjat, 

> You only need to check the flag when you set the serde to write data into topic.

Sorry, I don't follow. Is there a change you think needs making or is this good to be merged?

Thanks,

Andy

**Pull request merged**